### PR TITLE
Add option to read passphrase from a keyfile

### PR DESCRIPTION
**Pull request by oz123**
Similar to the same option with luks. Of course,
one can simply wrap bcachefs tool with a script and expect, but this is a nicer way of doing things.

It's the first time, I contribute C code to any project related to the kernel. Apologies, if I don't comply to standards.
I tried my best being a good neighbor. Please be easy on me.

See #210.

### CODE REVIEW

File: crypto.c [hunk 1]
```diff
@@ -75,6 +75,38 @@ char *read_passphrase_twice(const char *prompt)
...
+    if (buffer == NULL) {
+        fprintf(stderr, "Memory allocation failed\n");
+        fclose(file);
+        return NULL;
+    }
+
+    fread(buffer, 1, length, file);
+    buffer[length] = '\0'; // Null-terminate the string
+
+    // Close the file
```
Comment (koverstreet) [inline crypto.c:104]: Useless comment - delete this

File: crypto.c [hunk 2]
```diff
@@ -75,6 +75,38 @@ char *read_passphrase_twice(const char *prompt)
 	return pass;
 }
 
+char *read_password_from_file(const char *file_path) {
+    FILE *file;
+    char *buffer;
+    long length;
+
+    file = fopen(file_path, "r");
```
Comment (koverstreet) [inline crypto.c:83]: Declare it and initialize it on the same line

File: crypto.c [hunk 3]
```diff
@@ -75,6 +75,38 @@ char *read_passphrase_twice(const char *prompt)
 	return pass;
 }
 
+char *read_password_from_file(const char *file_path) {
```
Comment (koverstreet) [inline crypto.c:78]: We already have a helper for this - read_file_str() in tools-util.c

Comment (oz123) [reply crypto.c:78]: Nice. I'll use that! Thanks for your time.

Comment (oz123) [reply crypto.c]: I finally got the time to make the appropriate change.

**Comment by RlndVt** [comment]
Is it possible to also add this flag to the mount command? e.g. `bcachefs mount -f <keyFile> UUID=<uuid> <mountPoint>`

**Comment by oz123** [comment]
@RlndVt this looks like a good idea! I'll see if I can find some time for it next month. Unless, someone else wants to add it. I am pretty busy with `paid` work... Also, I don't know if Kent even wants to merge this PR.

**Comment by RlndVt** [comment]
Understandable. By the looks of it the mount command is written in rust. I have no experience with rust... yet...

**Comment by koverstreet** [comment]
one of my todo items was to make the mount command prompt for the passphrase when necessary - since we're now always mounting with the mount helper, it will help a lot with the keyring issues people have been having if the keyring is being used by the same process.

If you're looking at mount, could you take a look at that as well?

**Comment by RlndVt** [comment]
@koverstreet , if you want to pull this I can make an attempt at a draft PR for the mount command enhancement(s).

**Comment by RlndVt** [comment]
@koverstreet

I've been giving this some thought I was wondering if you had a plan how to handle fstab (and systemd).

A `-f keyfile` option might be the easiest to implement, but can't be entered or passed to fstab. Maybe with something similar to a `x-systemd-option`, or something similar but bcachefs specific. (`x-bcachefs-keyfile=<path>`).

Another alternative I thought of was using mount options. Probably a more complicated solution, but flexible enough to be `mount` and `fstab` agnostic. Don't know if it's even possible; the mount command will need to preprocess mount options to discover if a keyfile is specified.

Love to hear your thoughts.

**Pull request merged**